lsm-perf.py

```diff
@@ -0,0 +1,180 @@
+
+
+ROUNDS = 3
+REPETITIONS_PER_ROUND = 50
+WARMUP_RUNS = 5
+ON_VM_WORKLOAD_PATH = '~/lsm-perf-workload'
+QEMU_EXIT_CMD = b'\x01cq\n'  # -> ctrl+a c q
+
+
+def main(args):
+    try:
+        init_output_file(args.out)
+        for round in range(ROUNDS):
+            print('Starting round %d' % round)
+            for kernel in args.kernels:
+                results = evaluating_kernel(
+                    kernel_path=kernel.name,
+                    img_path=args.image.name,
+                    workload_path=args.workload.name,
+                    keyfile=args.key.name
+                )
+                write_results_to_file(args.out, kernel.name, round, results)
+    except KeyboardInterrupt:
+        print('\nExit prematurely')
+    finally:
+        args.out.close()
+    return 0
+
+
+def evaluating_kernel(kernel_path, img_path, workload_path, keyfile):
+    """Starts a VM with the kernel and evaluates its
+    performances on the provided workload
+    """
+    results = []
+    name = kernel_path[kernel_path.rfind('/') + 1:]
+    print_eta(name, info='connecting')
+
+    with VM(kernel_path, img_path, keyfile) as vm:
+        vm.scp_to(workload_path, ON_VM_WORKLOAD_PATH)
+        work_cmd = vm.ssh[ON_VM_WORKLOAD_PATH]
+
+        print_eta(name, info='Running warm up')
+        for _ in range(WARMUP_RUNS):
+            work_cmd()
+
+        for i in range(REPETITIONS_PER_ROUND):
+            results.append(int(work_cmd().strip()))
+            percentage = (i + 1) * 100 / REPETITIONS_PER_ROUND
+            print_eta(name, info='%d%%' % percentage)
+
+        vm.ssh.path(ON_VM_WORKLOAD_PATH).delete()
+
+    stats = ('\taverage=%d, stdev=%d' %
+             (statistics.mean(results), statistics.stdev(results)))
+    print_eta(name, info=stats)
+    print()
+    return results
+
+
+class VM:
+    """Represents a qemu VM with an SSH connection.
+    To use with the `with` construct
+    """
+    def __init__(self, kernel_path, img_path, keyfile):
+        """Starts the qemu VM"""
+        qemu_args = construct_qemu_args(kernel_path, img_path)
+        self.process = local['qemu-system-x86_64'].popen(qemu_args)
+        self.ssh = None
+        self.key = keyfile
+
+    def __enter__(self):
+        """Initialize ssh connection"""
+        c = 5
+        while self.ssh is None:
+            time.sleep(1)
+            try:
+                self.ssh = SshMachine(
+                    '127.0.0.1', user='root', port=5555, keyfile=self.key)
+            except (EOFError, plumbum.machines.session.SSHCommsError) as e:
+                c -= 1
+                if c == 0:
+                    raise e
+        return self
+
+    def __exit__(self, type, value, traceback):
+        """Stops the SSH connection and the VM"""
+        if self.ssh is not None:
+            self.ssh.close()
+            self.ssh = None
+        self.process.stdin.write(QEMU_EXIT_CMD)
+        self.process.terminate()
+        time.sleep(0.5)
```

Comment:
I'll generally ask for a comment explaining why a `sleep` is here - what are we wating for?

 The one in the retry loop above was obvious enough that it isn't necessary but this one ought to be explained IMO.

Reply:
I removed this one (see comment above), but I will explain it better if I need other `sleep`s.